Conda and pixi notebooks fall back to UV pooled kernels — packages installed via UV, not conda/pixi

## Bug

Conda and pixi notebooks fall back to UV pooled kernels — packages installed via UV, not conda/pixi.

## Root causes (two stacked bugs)

### 1. MCP routing: unscoped `env_source: "auto"` (fixed in #1606)

`create_notebook` and `restart_kernel` sent `env_source: "auto"` to the daemon, which resolved using the system default (`default_python_env`, typically `uv`). Fixed by sending scoped `"auto:pixi"` / `"auto:conda"` / `"auto:uv"` based on notebook metadata.

### 2. Conda warmup timeout: 30s too short (open — prewarm branch)

Even with correct routing, the conda pool was in backoff because the warmup script times out:

```
2026-04-07 00:44:16 [ERROR] Conda warmup timed out
2026-04-07 00:44:47 [WARN]  Conda pool in backoff: 9 consecutive failures 
  (Conda warmup script failed (ipykernel may not be installed)), waiting 300s
```

**The error message is misleading.** ipykernel IS in the solver specs and IS installed. The actual failure is a **30-second timeout** in the warmup script (`daemon.rs:2951`). The warmup imports ALL packages — ipykernel, IPython, ipywidgets, anywidget, nbformat, traitlets, zmq, matplotlib, seaborn, numpy, pandas, polars — and first-time imports in a fresh conda env take 34+ seconds due to .pyc compilation, C extension loading, and BLAS discovery.

**Evidence**: timing the warmup script on an existing conda env (with .pyc already compiled):
```
$ time python -c "import ipykernel; import IPython; ... import matplotlib; import pandas; ..."
warmup complete
real    0m34.298s   ← exceeds 30s timeout even with warm caches
```

### Failure cascade

1. Daemon starts, creates conda env (solver succeeds, install succeeds)
2. Warmup script runs → times out at 30s → returns `false`
3. `false` triggers generic error "ipykernel may not be installed" (misleading)
4. Pool enters exponential backoff: 60s → 120s → 240s → 300s (capped)
5. Existing pre-warmed envs from previous session get used up
6. New conda notebook requests fall back to UV prewarmed pool

### Pixi

No pixi warmup log entries observed. Pixi pool shows 2/2 ready but may have same timeout issue. Needs investigation.

## Fixes

### Done
- **#1606**: MCP routing sends scoped `env_source` (merged)

### Needed (prewarm branch)
- **Increase warmup timeout** from 30s to 120s (both conda and UV paths in `daemon.rs:2951` and `daemon.rs:3344`)
- **Fix error message**: the timeout branch at `daemon.rs:2981` returns `false` which triggers the generic "ipykernel may not be installed" at line 2934. Should propagate the actual error ("warmup timed out")
- **Better warmup approach**: use try/except per import so one slow package doesn't kill the whole warmup. The [`prewarm` PyPI package](https://pypi.org/project/prewarm/) (owned by @rgbkrk) does exactly this — separate validation (is ipykernel importable?) from cache warming (import everything with try/except)
- **Consider two-phase warmup**: fast phase (verify ipykernel works, add to pool) then background slow phase (warm caches for matplotlib etc.)

## Relevant code

| Location | What |
|----------|------|
| `daemon.rs:265` | `build_python_warmup_script` — generates the import script |
| `daemon.rs:2698-2706` | Conda solver specs (ipykernel IS included) |
| `daemon.rs:2949-2985` | Conda warmup with 30s timeout |
| `daemon.rs:3341-3350` | UV warmup with same 30s timeout |
| `daemon.rs:2926-2937` | Misleading "ipykernel may not be installed" error |

## Environment

- nteract nightly 2.1.1-nightly.202604070640
- macOS arm64, conda-forge
- Daemon pool: UV 3/3 ready, Conda 0/3 (backoff), Pixi 2/2 ready

--- a/crates/runt-mcp/src/tools/kernel.rs
+++ b/crates/runt-mcp/src/tools/kernel.rs
@@ -63,10 +63,11 @@ pub async fn restart_kernel(
         tracing::warn!("confirm_sync failed before restart_kernel launch: {e}");
     }
 
-    // Step 2: Determine kernel type and env_source from RuntimeState.
-    // Use scoped auto-detect (auto:uv, auto:conda, auto:pixi) to stay within
-    // the original package manager family while re-checking metadata for new deps.
-    // This matches the Python bindings' restart logic in session_core.rs.
+    // Step 2: Determine kernel type and env_source.
+    // Use metadata-based detection (not RuntimeState env_source) to scope the
+    // auto-detect. This ensures the correct package manager pool is used even
+    // if the previous kernel was launched with a different env (e.g., UV default
+    // when the notebook metadata says pixi). See #1605.
     let (kernel_type, env_source) = {
         let state = handle.get_runtime_state().ok();
         let kernel_type = state
@@ -80,12 +81,12 @@ pub async fn restart_kernel(
                 }
             })
             .unwrap_or_else(|| "python".to_string());
-        let env_source = match state.as_ref().map(|s| s.kernel.env_source.as_str()) {
-            Some("uv:prewarmed") => "auto:uv".to_string(),
-            Some("conda:prewarmed") => "auto:conda".to_string(),
-            Some("pixi:prewarmed") => "auto:pixi".to_string(),
-            Some("") | None => "auto".to_string(),
-            Some(s) => s.to_string(),
+        // Scope auto-detect based on notebook metadata, not stale env_source
+        let detected_manager = super::deps::detect_package_manager(&handle);
+        let env_source = match detected_manager.as_str() {
+            "pixi" => "auto:pixi".to_string(),
+            "conda" => "auto:conda".to_string(),
+            _ => "auto:uv".to_string(),
         };
         (kernel_type, env_source)
     };

--- a/crates/runt-mcp/src/tools/session.rs
+++ b/crates/runt-mcp/src/tools/session.rs
@@ -452,7 +452,15 @@ pub async fn create_notebook(
                         tracing::warn!("confirm_sync failed before create_notebook relaunch: {e}");
                     }
 
-                    // Shutdown and relaunch with auto-detect (daemon reads deps from metadata)
+                    // Shutdown and relaunch with scoped auto-detect so the daemon
+                    // uses the correct package manager pool (not the system default).
+                    // "auto:pixi" → pixi pool/inline, "auto:conda" → conda pool/inline,
+                    // "auto" → follows default_python_env (which may differ from requested).
+                    let scoped_env_source = match pkg_manager {
+                        "pixi" => "auto:pixi",
+                        "conda" => "auto:conda",
+                        _ => "auto:uv",
+                    };
                     let _ = s
                         .handle
                         .send_request(NotebookRequest::ShutdownKernel {})
@@ -462,7 +470,7 @@ pub async fn create_notebook(
                         .handle
                         .send_request(NotebookRequest::LaunchKernel {
                             kernel_type: runtime.to_string(),
-                            env_source: "auto".to_string(),
+                            env_source: scoped_env_source.to_string(),
                             notebook_path: None,
                         })
                         .await;

